contracts/secretswap_router/src/contract.rs

```diff
@@ -0,0 +1,378 @@
+                )?);
+            } else if first_hop.from_token.native_denom == Some("uscrt".into()) {
+                // first hop is SCRT
+                msgs.push(
+                        // build swap msg for the next hop                    CosmosMsg::Wasm(WasmMsg::Execute {
+                        contract_addr: first_hop.pair_address,
+                        callback_code_hash: first_hop.pair_code_hash,
+                        msg: to_binary(&NativeSwap::Swap {
+                            offer_asset: Asset {
+                                amount,
+                                info: AssetInfo::NativeToken {
+                                    denom: "uscrt".into(),
+                                },
+                            },
+                            // set expected_return to None because we don't care about slippage mid-route
+                            expected_return: None,
+                            // set the recepient of the swap to be this contract (the router)
+                            to: Some(env.contract.address.clone()),
+                        })?,
+                        send: vec![Coin::new(amount.u128(), "uscrt")],
+                    }),
+                );
+            } else {
+                // shouldn't be here because we've tested this a few lines before
+                return Err(StdError::generic_err(
+                    "cannot build swap message for first hop",
+                ));
+            }
+
+            msgs.push(
+                // finalize the route at the end, to make sure the route was completed successfully
+                CosmosMsg::Wasm(WasmMsg::Execute {
+                    contract_addr: env.contract.address.clone(),
+                    callback_code_hash: env.contract_code_hash.clone(),
+                    msg: to_binary(&HandleMsg::FinalizeRoute {})?,
+                    send: vec![],
+                }),
+            );
+
+            Ok(HandleResponse {
+                messages: msgs,
+                log: vec![],
+                data: None,
+            })
+        }
+        HandleMsg::Receive {
+            from: _,
+            msg: None,
+            amount,
+        } => {
+            // This is a receive msg somewhere along the route
+            // 1. load route from state (Y/Z -> Z/W)
+            // 2. save the remaining route to state (Z/W)
+            // 3. send `amount` Y to pair Y/Z
+
+            // 1'. load route from state (Z/W)
+            // 2'. this is the last hop so delete the entire route state
+            // 3'. send `amount` Z to pair Z/W with recepient `to`
+            match read_route_state(&deps.storage)? {
+                Some(RouteState {
+                    is_done: _,
+                    remaining_route,
+                }) => {
+                    if remaining_route.hops.len() == 0 {
+                        return Err(StdError::generic_err("route must be at least 1 hop"));
+                    }
```

Comment:
Why not just let this fail if this can never happen normally?

Reply:
This if clause is there mainly to avoid getting a weird panic when we access `hops[0]` 2 lines after this, and return a clear error message to the user.

https://github.com/enigmampc/SecretSwap/blob/364b785a27e5ba451cfd3d4eb2dfa7c9af486b59/contracts/secretswap_router/src/contract.rs#L181

Let me know what you think.